feat: cookieless embed version 2 (#106)
looker-open-source/embed-sdk: README.md

## Introduction

The Looker JavaScript Embed SDK is designed to facilitate using Looker embedded content in your web application. The goal is to make communication between a host website and one or more embedded dashboards, looks, explores and extensions easier and more reliable.

> The Looker JavaScript Embed SDK typically uses embed SSO to sign an embed url in order to authenticate the user of the embed. This mechanism relies on Looker cookies being available to the embedded IFRAME in order for the application to work. Looker also provides a mechanism that allows embedded Looker IFRAMES to work without the need for cookies. Details can be found [here](#cookieless). Embed SDK functionality that will not work with cookieless embed are identified in this document.

Reviewer: ```suggestion
The Looker JavaScript Embed SDK typically uses embed SSO to sign an embed url in order to authenticate the user of the embed. This mechanism relies on Looker cookies being available to the embedded IFRAME in order for the application to work. Looker also provides a mechanism that allows embedded Looker IFRAMES to work without the need for cookies. Details can be found [here](#cookieless). Embed SDK functionality that will not work with cookieless embed is identified in this document.
```

/embed/looks/4?embed_domain=https://mywebsite.com&sdk=2
```

The SDK cannot add this parameter itself because it part of the signed SSO URL.

## The Auth Endpoint

This section does not apply to cookieless embed as an alternate mechanism for authentication is used. See the [cookieless embed](#cookieless) section for details.

In order to use the embed SDK on the frontend you must supply a backend service that handles authentication. This service is called by the SDK to generate a signed iframe URL that is unique to the requesting user. The backend process can either generate the signed embed URL itself using an embed secret or the backend process can generate the URL by calling the Looker API. Manual URL generation and signing avoids calling the Looker API resulting in decreased latency. Calling the Looker API requires less code and can be easier to maintain.

### Backend Process

This section does not apply to cookieless embed but a backend process is still required. See the [cookieless embed](#cookieless) section for details.

The _backend_ process entails hosting a service at an endpoint such as `/auth` which does the following:

> 1. The backend service initializes the [Looker API SDK](https://docs.looker.com/reference/api-and-integration/api-sdk) based on a client API key and secret typically stored in `Looker.ini` file.

Reviewer: @kenvlooker is there a direct link to the new docs home we should use instead here?

Reviewer 2: Good catch, we could update the link to this URL: https://cloud.google.com/looker/docs/api-sdk

Reviewer 2: That said, the existing redirect is working.

Author: Updated

  first_name?: string
  last_name?: string
  session_length: number
  force_logout_login?: boolean
  permissions: LookerUserPermission[]
  models: string[]
  group_ids?: number[]
  external_group_id?: string
  user_attributes?: { [key: string]: any }
  access_filters: { [key: string]: any }
}
```

<a name='cookieless' id='cookieless'></a>

## Cookieless Embed

> Note that cookieless embed does not yet support the use of `withUrl`. An attempt to do use this functionality will result in an error being thrown.

Reviewer: ```suggestion
Note that cookieless embed does not yet support the use of `withUrl`. An attempt to use this functionality will result in an error being thrown.
```

> Looker cookieless embed allows the Looker application to be embedded by an html page that is served from a different domain than the Looker host. With SSO embed, to avoid third party cookie blocking, the Looker application must be served from a sub domain of the hosting application OR the user must enable third party cookies in the browser. Enabling cookieless embedding is documented in more detail [here](https://cloud.google.com/looker/docs/r/sdk/cookieless-embed). Cookieless embed is available with Looker version 22.18 and above.

Reviewer 2: ```suggestion
Looker cookieless embed allows the Looker application to be embedded by an html page that is served from a different domain than the Looker host. With SSO embed, to avoid third party cookie blocking, the Looker application must be served from a sub domain of the hosting application OR the user must enable third party cookies in the browser. Enabling cookieless embedding is documented in more detail [here](https://cloud.google.com/looker/docs/r/sdk/cookieless-embed). Cookieless embed is available with Looker version 22.20 and above.
```

Cookieless embed works by using short lived tokens that are kept in the browser and are used to reference the actual session in the Looker server. The Looker UI keeps track of the tokens, and before they expire, requests that the hosting application generate new ones. To this end, the host application is required to implement functionality in the client and in the server.

This functionality will:

- Acquire a new session either by creating or attaching to an existing session associated with the browser (this allows the user to create new IFRAMEs and use the same session).
- Generate new tokens.

### Acquire session backend process

> This process will be called everytime a Looker embed IFRAME is created. The acquire session backend process requires that the Looker api endpoint `acquire_embed_cookieless_session` be called to create an embed session or attach to an existing embed session. This endpoint is responsible for creating the embed user. This is different from embed SSO whereby the embed login creates the embed user. One major difference in the payloads is that `force_logout_login` is ignored by `acquire_embed_cookieless_session`. Cookieless embed logins ALWAYS force logout login (as there should be no Looker cookies this should be a noop). It should also be noted that this endpoint will NOT update the embed user in Looker if it already exists. There are other endpoints in Looker that can do this.

Reviewer: ```suggestion
This process will be called every time a Looker embed IFRAME is created. The acquire session backend process requires that the Looker api endpoint `acquire_embed_cookieless_session` be called to create an embed session or attach to an existing embed session. This endpoint is responsible for creating the embed user. This is different from embed SSO whereby the embed login creates the embed user. One major difference in the payloads is that `force_logout_login` is ignored by `acquire_embed_cookieless_session`. Cookieless embed logins ALWAYS force logout login (as there should be no Looker cookies this should be a noop). It should also be noted that this endpoint will NOT update the embed user in Looker if it already exists. There are other endpoints in Looker that can do this.
```

    const { api_url, client_id, client_secret, verify_ssl } = config
    try {
      const lookerSettings = DefaultSettings()
      lookerSettings.readConfig = () => {
        return {
          client_id,
          client_secret,
        }
      }
      lookerSettings.base_url = api_url
      lookerSettings.verify_ssl = verify_ssl
      lookerSession = new NodeSession(lookerSettings)
      lookerSession.login()
    } catch (error) {
      console.error('login failed', { error })
      throw error
    }
  }
}

// Simple function to acquire the embed session.
// Note as an additional layer of security the user agent of users
// browser is associated with the embed session. It is important
// that this is available when the embed session is created.
const acquireEmbedSessionInternal = async (userAgent, user) => {
  try {
    const cacheKey = `${user.external_user_id}/${userAgent}`
    const embedSession = embedSessions[cacheKey]
    const request = {
      ...user,
      session_reference_token: embedSession?.session_reference_token,
    }
    const sdk = new Looker40SDK(lookerSession)
    const response = await sdk.ok(

>       sdk.acquire_embed_cookieless_session(request, {

Reviewer: Do you think it's worth using functional SDK syntax here instead?

Author: As per discussion, has no benefit in node and is more discoverable. Leaving as is for now

The example below uses javascript, but you may utilize the Looker SDK in the language of your choice. Note that this is very simplistic implementation for demonstration purposes only. An actual implementation should be a lot more robust.

```javascript
app.get('/generate-embed-tokens', async function (req, res) {
  try {
    const tokens = await generateEmbedTokens(req.headers['user-agent'], user)
    res.json(tokens)
  } catch (err) {
    res.status(400).send({ message: err.message })
  }
})

export async function generateEmbedTokens(userAgent, user) {
  const cacheKey = `${user.external_user_id}/${userAgent}`
  const embedSession = embedSessions[cacheKey]
  if (!embedSession) {
    console.error(
      'embed session generate tokens failed, session not yet acquired'
    )
    throw new Error(
      'embed session generate tokens failed, session not yet acquired'
    )
  }
  await acquireLookerSession()
  try {
    const { api_token, navigation_token, session_reference_token } =
      embedSession
    const sdk = new Looker40SDK(lookerSession)
    const response = await sdk.ok(

>       sdk.generate_tokens_for_cookieless_session(

Reviewer: FunSDK here too?

Author: See  previous comment

    throw new Error(
      `generate-embed-tokens failed: ${resp.status} ${resp.statusText}`
    )
  }
  return await resp.json()
}
```

### Initializing the Looker SDK in the frontend

Cookieless embed is initialized by calling `LookerEmbedSDK.initCookieless` passing in the Looker host value and the two callbacks described previously. Once a Looker embed IFRAME is created it will communicate with the Embed SDK running in the host application and use the callbacks appropriately.

```javascript

const acquireEmbedSessionCallback = async () => {
  . . .
}

const generateEmbedTokensCallback = async () => {
  . . .
}

LookerEmbedSDK.initCookieless(
  'looker.example.com',
  acquireEmbedSessionCallback,
  generateEmbedTokensCallback
)
```

### Loading

## Demo

A simple demo is provided in the `/demo` directory that uses a basic JS frontend and a Python backend. The example backend `demo.py` uses the Looker API to create a signed URL. The example backend `demo_self_signed.py` uses the embed secret and a helper function to sign the URL. The instructions below are for the example using the Looker API.

### Step 1 - Enable Embedding in your Looker instance

> Enabling SSO embedding is documented in more detail [here](https://docs.looker.com/r/sdk/sso-embed).

Reviewer: @kenvlooker recommendation for new docs home link?

Reviewer 2: That URL is already redirected and should work fine, but if we want to update it, the current URL for that page is https://cloud.google.com/looker/docs/single-sign-on-embedding.

Author: Updated

Enabling cookieless embedding is documented in more detail [here](https://cloud.google.com/looker/docs/r/sdk/cookieless-embed).

- Navigate to Admin > _Platform_ Embed on your Looker instance. This requires Admin privileges.
- The demo server runs by default at [http://localhost:8080](http://localhost:8080). By adding that address to "Embedded Domain Whitelist" you can enabled the demo to receive messages from Looker.

> - Turn on "Embed Authentication"

Author: I'm going to accept this for now as what you have is currently correct as far as the UI is concerned. This will probably need to change though. Embed SSO Authentication also turns on cookieless embed and cookieless embed is not SSO, at least not as far as I am concerned.

- In order to use embedding you must generate an "Embed Secret" for SSO embedding and/or a JWT secret for cookieless embedding. Note that a Looker instance can support both types of embedding at the same time.

Additional steps for cookieless embed:

- Navigate to Admin > _Platform_ Embed on your Looker instance. This requires Admin privileges.
- Generate an Embed JWT secret. This is used internally and the hosting application does not need to know what it is.
- Navigate to Admin > _Labs_ Experimental on your Looker instance. This requires Admin privileges.
- Toggle "Cookieless Embed" on. Note that a Looker instance can support SSO and Cookieless embed clients.

### Step 2 - Customize the Demo settings for your Looker instance

Note that `demo.py` and `demo_self_signed.py` have NOT been updated to support cookieless embedding. This section ONLY applies to SSO embedding. The cookieless embed demo currently requires the use of the development server.

> - If you are using the main `demo.py`, provide your API credentials to the server by updating `demo/looker.ini` following [these instructions](https://community.looker.com/technical-tips-tricks-1021/the-how-to-on-initializing-the-sdk-with-different-profiles-in-your-ini-file-26846), with credentials obtained from [the Users page](https://docs.looker.com/reference/api-and-integration/api-auth).

Reviewer: @kenvlooker another link update (or two?)

Reviewer 2: The community link is correct. The redirect from the docs.looker.com URL is in place, so that link is fine, too. That said, if we want to replace it with the current URL, that page is at https://cloud.google.com/looker/docs/api-auth.

Author: Updated
